### PR TITLE
Cache more things in HandlerMethodDescriptor

### PR DESCRIPTION
**Pull request by pranavkm**
Add tests for DefaultPageHandlerMethodSelector

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageHandlerMethodSelector.cs [hunk 1]
```diff
@@ -3,140 +3,106 @@
...
-            public static HandlerMethodAndMetadata Create(HandlerMethodDescriptor handler)
-            {
-                var name = handler.Method.Name;
+            var handlers = context.ActionDescriptor.HandlerMethods;
+            List<HandlerMethodDescriptor> handlersToConsider = null;
 
-                string httpMethod;
-                if (name.StartsWith("OnGet", StringComparison.Ordinal))
+            var formAction = Convert.ToString(context.RouteData.Values[FormAction]);
+            for (var i = 0; i < handlers.Count; i++)
```
Comment (ryanbrandenburg) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageHandlerMethodSelector.cs]: More of a question than feedback, you seem to prefer `for` over `foreach` even when you only use the `i` to get elements, is that just personal preference or something else?

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageHandlerMethodSelector.cs [hunk 2]
```diff
@@ -3,140 +3,106 @@
...
-                    : name.Length - formactionStart;
+                handlersToConsider.Add(handler);
+            }
 
-                var formaction = formactionLength == 0 ? null : name.Substring(formactionStart, formactionLength);
+            return handlersToConsider;
+        }
 
-                return new HandlerMethodAndMetadata(handler, httpMethod, formaction);
+        private static int GetScore(HandlerMethodDescriptor descriptor)
```
Comment (ryanbrandenburg) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageHandlerMethodSelector.cs]: Could/should we include a comment either here or in the calling function giving a little description about what exactly these scores mean and how they're used? It feels a tad opaque to me.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageHandlerMethodSelector.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageHandlerMethodSelector.cs]: http://stackoverflow.com/questions/18552669/memory-allocation-when-using-foreach-loops-in-c-sharp

They add up and there's more noise for profiling. I just got in the habit and it looks like it's spreading. You won't see me complain if you use `foreach` though unless we know it to be in a super hot path.

I believe when we did a pass to remove Linq on per-request code paths we removed something like 5% of all the allocations for an API scenario. It's not a huge thing but it was definitely worth doing.

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/HandlerMethodDescriptor.cs [hunk 3]
```diff
@@ -12,5 +13,9 @@ public class HandlerMethodDescriptor
         public MethodInfo Method { get; set; }
 
         public Func<Page, object, Task<IActionResult>> Executor { get; set; }
+
+        public string HttpMethod { get; set; }
+
+        public StringSegment FormAction { get; set; }
```
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Infrastructure/HandlerMethodDescriptor.cs]: Why not just a string here?

File: src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageHandlerMethodSelector.cs (hunk 2 above)
Comment (rynowak) [inline src/Microsoft.AspNetCore.Mvc.RazorPages/Internal/DefaultPageHandlerMethodSelector.cs]: Yeah this is such @rynowak code that it deffo needs a comment.

**Pull request closed without merging**